pkg/storage/mvcc.go

```diff
@@ -2128,28 +2129,46 @@ func MVCCClearTimeRange(
 	//
 	// Intents must be seen so that an error can be returned. We never delete an
 	// intent here.
-	it := rw.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{LowerBound: key, UpperBound: endKey})
-	defer it.Close()
+	iterOptions := io
+	iterOptions.LowerBound = key
+	iterOptions.UpperBound = endKey
+	var iter MVCCIterator
+	var timeBoundIter MVCCIterator
+	if !iterOptions.MinTimestampHint.IsEmpty() && !iterOptions.MaxTimestampHint.IsEmpty() {
+		timeBoundIter = rw.NewMVCCIterator(MVCCKeyIterKind, iterOptions)
+	}
+	iter = rw.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{LowerBound: key,
+		UpperBound: endKey})
+	defer iter.Close()
 
-	var clearedMetaKey MVCCKey
-	var clearedMeta enginepb.MVCCMetadata
-	var restoredMeta enginepb.MVCCMetadata
-	for it.SeekGE(MVCCKey{Key: key}); ; it.Next() {
-		ok, err := it.Valid()
+	// Use a TBI to find the start key if TBI optimization is enabled.
+	if timeBoundIter != nil {
+		timeBoundIter.SeekGE(MVCCKey{Key: key})
+		ok, err := timeBoundIter.Valid()
 		if err != nil {
 			return nil, err
 		} else if !ok {
-			break
+			return nil, nil
 		}
+		iter.SeekGE(timeBoundIter.Key())
+	} else {
+		iter.SeekGE(MVCCKey{Key: key})
+	}
 
```

Comment:
I might say go ahead and do valid/err check here, close to the Seek where a reader would expect it, and then switch the `for` loop below to single condition test instead of init/test/step triple?

 or maybe even just `for {` and do the Valid check and return / break paired with the Next/Seek calls at the end, so we don't need this mutable `err`  ? I donno.